src/messenger/messenger.cpp

```diff
@@ -93,80 +93,31 @@
-   */
-  static ZmqMessage Build(uint64_t send_msg_id, uint64_t recv_cb_id, const std::string &sender_id,
-                          std::string_view message) {
-    return ZmqMessage{sender_id, fmt::format("{}-{}-{}", send_msg_id, recv_cb_id, message)};
-  }
-
-  /**
-   * Parse the given payload into a ZmqMessage.
-   * @param routing_id      The message's routing ID.
-   * @param message         The message received.
-   * @return A ZmqMessage encapsulating the given message.
-   */
-  static ZmqMessage Parse(const std::string &routing_id, const std::string &message) {
-    return ZmqMessage{routing_id, message};
-  }
-
-  /** @return The ID of this message (sender side). */
-  uint64_t GetMessageIdSender() const { return send_msg_id_; }
-
-  /** @return The ID of this message (receiver side). */
-  uint64_t GetCallbackIdReceiver() const { return recv_cb_id_; }
-
-  /** @return The routing ID of this message. */
-  std::string_view GetRoutingId() const { return std::string_view(routing_id_); }
-
-  /** @return The message itself. */
-  std::string_view GetMessage() const { return message_; }
+ZmqMessage ZmqMessage::Build(uint64_t source_cb_id, uint64_t dest_cb_id, const std::string &routing_id,
+                             std::string_view message) {
+  return ZmqMessage{routing_id, fmt::format("{}-{}-{}", source_cb_id, dest_cb_id, message)};
+}
 
-  /** @return The raw payload of the message. */
-  std::string_view GetRawPayload() const { return std::string_view(payload_); }
+ZmqMessage ZmqMessage::Parse(const std::string &routing_id, const std::string &message) {
+  return ZmqMessage{routing_id, message};
+}
 
- private:
-  friend Messenger;
-
-  /** Construct a new ZmqMessage with the given routing ID and payload. Payload of form ID-MESSAGE. */
-  ZmqMessage(std::string routing_id, std::string payload)
-      : routing_id_(std::move(routing_id)), payload_(std::move(payload)), message_(payload_) {
-    if (payload_.empty()) {
-      send_msg_id_ = 0;
-      recv_cb_id_ = 0;
-    } else {
-      // TODO(WAN): atoi, stoull, from_chars, etc? Error checking in general.
-      UNUSED_ATTRIBUTE int check =
-          std::sscanf(payload_.c_str(), "%" SCNu64 "-%" SCNu64 "-", &send_msg_id_, &recv_cb_id_);
-      NOISEPAGE_ASSERT(2 == check, "Couldn't parse the message header.");
-
-      // Remove the prefix up to the second '-'
-      message_.remove_prefix(message_.find_first_of('-') + 1);
-      message_.remove_prefix(message_.find_first_of('-') + 1);
-    }
+ZmqMessage::ZmqMessage(std::string routing_id, std::string payload)
+    : routing_id_(std::move(routing_id)), payload_(std::move(payload)), message_(payload_) {
+  if (payload_.empty()) {
+    source_cb_id_ = 0;
+    dest_cb_id_ = 0;
+  } else {
+    // TODO(WAN): atoi, stoull, from_chars, etc? Error checking in general.
+    UNUSED_ATTRIBUTE int check =
+        std::sscanf(payload_.c_str(), "%" SCNu64 "-%" SCNu64 "-", &source_cb_id_, &dest_cb_id_);
```

Comment:
This is my opinion, so feel free to ignore it if you disagree. I guess this was already here, but it makes me a little nervous seeing the expected format of the message hard coded into this constructor here and on lines 117, 118. It might be more obvious to make these constants and refer to those.

Reply:
unfortunately it is also hardcoded into the model server message format, off in Python land, if I'm not mistaken. but yea, I can try to refactor -- I seem to recall facing some weird issue which ended with those SCNu64's there. you're right to point out hacks, please keep doing so!